### DISCUSSION TITLE
Speed

### DISCUSSION THREAD

**Discussion by john9631** [ghostty-org/ghostty#4032] [Q&A]
I watched your video on speed with interest.  You measure speed in one way but there is another: responsiveness.

Typometer, https://github.com/pavelfatin/typometer, measures this simple latency and although ghostty is quite nice you'll find it slow compared to xterm and st.  Any thoughts.

Regards

**Comment by dm17** [comment]
Yes! Would be very interested in whether the freshness of this terminal emulator rewrite can give us an edge in the latency wars.
Here are some relevant benchmarks:
https://github.com/moktavizen/terminal-benchmark
https://thume.ca/2020/05/20/making-a-latency-tester/
https://nick-black.com/dankwiki/index.php?title=Notcurses#Terminal_emulators

**Reply by q3cpma** [reply]
First section of the first one (input latency) is especially important in my eyes.

**Reply by dm17** [reply]
Yes one would think latency would be something they'd want to showcase for zig

**Reply by rhodes-b** [reply]
I'm not sure why people get any idea that using Zig for the language matters at all, it has very little relevance on showcasing it.

Zig is a systems language that is compiled with no garbage collection that uses LLVM (currently), as such its automatically going to be in a similar class to C / C++ / Rust.

I could write some absolute slop that makes zig perform worse than python that doesn't mean zig as a language is bad, it means I wrote horrible code. 

Trying to benchmark the performance of zig through ghostty by comparing it to another terminal written in another language is totally incomparable.

For example to say ghostty is faster than wezterm in X benchmark that must mean zig is better then rust makes no sense, when they have extremely different architectures of the underlying program and support different things.

Now is ghostty trying to be fast absolutely its listed as one of the [main goals](https://ghostty.org/docs/about#fast). Does using Zig help that again absolutely, but whenever mitchell gets asked why zig its simply because he likes zig not because its more performant than language X and not to spread to spread zig propaganda.

Some areas of ghostty are also known to not have the best performance currently like images (kitty image protocol) ghostty is much slower then kitty in that case, but that again isn't any fault of zig there just hasn't been any time spent looking where the slowdowns are etc

**Reply by dm17** [reply]
Just to be clear, I don't think it has anything to do with benchmarking zig. It isn't a language war thing, but I can see how what I wrote might imply that... But as you state on 'main goals' there are many types of performance. For some reason some people really like having at least one terminal installed which they know has the lowest latency (which if I'm not mistaken is either alacritty or xterm - or kitty, if we change the defaults). Of course which benchmark published so far that we can trust is another story. Thanks for explaining a bit more!

**Reply by john9631** [reply]
In my experience the lowest latency is likely to be achieved with either xterm or st (modified version).  Anything adding buffering tends to add delay from character in to out.

**Reply by dm17** [reply]
True but we don't know that has to always be the case. As Mr Hashimoto says, there's low hanging fruit in these areas of optimization which haven't been a priority yet. I remain optimistic that I'll have the best of both worlds with ghostty eventually.
